Document safe Workflow patch retirement (#257)
temporalio/skill-temporal-developer: references/go/versioning.md

Author: Do we need to provide the same treatment to the other language SDK `versioning.md`s?

- `maxSupported`: current/newest version
- Returns `maxSupported` for new executions; returns the recorded version on replay

### Retirement Lifecycle

> Treat `GetVersion` cleanup as replay-safety work, not ordinary dead-code removal. Removing an old branch and deleting the first marker call are separate changes with different gates.

Reviewer: ```suggestion
`GetVersion` requires care to remain replay safe. Removing an old branch and deleting the first marker call are separate changes with different gates.
```

}
```

For new executions, `GetVersion` returns `1` and records a marker. For replay of pre-change workflows with no marker for `Step1`, it returns `workflow.DefaultVersion` (`-1`). **`DefaultVersion` is not version `0`**; do not search for `Step1-0` unless the code actually recorded version `0`.

> #### Step 2: Inventory and prove the old version has left retention

Reviewer: This section got really big in Go for some reason, but not the others (the smaller versions seem preferable. There's a lot of redundancy with the non-language-specific file).
